### PR TITLE
wp-cli-comform command

### PR DESCRIPTION
**Pull request by szepeviktor**
Adding line comments...

### CODE REVIEW

File: includes/wp-cli.php [hunk 1]
```diff
@@ -1,6 +1,6 @@
 <?php
 /**
- * Stream commands for WP-CLI
+ * Stream command for WP-CLI
```
Comment (szepeviktor) [inline includes/wp-cli.php]: only one

File: includes/wp-cli.php [hunk 2]
```diff
@@ -8,96 +8,152 @@
...
+	 * * meta.score
+	 * * meta.sort
+	 *
+	 * ## EXAMPLES
+	 *
+	 *     wp stream query --author_role__not_in=administrator --date_after=2015-01-01T12:00:00
+	 *     wp stream query --author=1 --action=login --records_per_page=50 --fields=created
 	 *
 	 * @see WP_Stream_Query
-	 * @subcommand query
```
Comment (szepeviktor) [inline includes/wp-cli.php]: redundant

File: includes/wp-cli.php [hunk 3]
```diff
@@ -8,96 +8,152 @@
...
+	 * @see https://github.com/wp-stream/stream/wiki/Query-Reference
 	 */
 	public function query( $args, $assoc_args ) {
-		self::connection();
```
Comment (szepeviktor) [inline includes/wp-cli.php]: no need to be static

File: includes/wp-cli.php [hunk 4]
```diff
@@ -8,96 +8,152 @@
...
+		$start             = microtime( true );
+		$query_args        = array();
+		$formatted_records = array();
 
-		$start      = microtime( true );
-		$query_args = array();
```
Comment (szepeviktor) [inline includes/wp-cli.php]: removed fany timer, a wp-cli user is advanced enough to know what is happening

File: includes/wp-cli.php [hunk 5]
```diff
@@ -8,96 +8,152 @@
...
+		$this->connection();
 
-		foreach ( $assoc_args as $key => $value ) {
-			$query_args[ $key ] = $value;
+		if ( empty( $assoc_args['fields'] ) ) {
+			$fields = array( 'created', 'ip', 'author', 'author_meta.user_login', 'author_role', 'summary' );
```
Comment (szepeviktor) [inline includes/wp-cli.php]: Where to put the YAML how-to?

File: includes/wp-cli.php [hunk 6]
```diff
@@ -8,96 +8,152 @@
...
 
-		if ( empty( $query_args['fields'] ) ) {
-			$defaults = array( 'created', 'ip', 'author', 'author_meta.user_login', 'author_role', 'summary' );
-
-			/**
-			 * Filter for default fields when arg is not provided
-			 *
-			 * @return array $defaults
-			 */
-			$defaults = apply_filters( 'wp_stream_wp_cli_default_fields', (array) $defaults );
```
Comment (szepeviktor) [inline includes/wp-cli.php]: removed filtering

File: includes/wp-cli.php [hunk 7]
```diff
@@ -8,96 +8,152 @@
...
-
-			$query_args['fields'] = implode( ',', array_map( 'trim', $defaults ) );
-		}
+		$query_args['fields'] = implode( ',', $fields );
 
 		$records = wp_stream_query( $query_args );
 
-		if ( empty( $records ) ) {
-			WP_CLI::success( __( 'No results found.', 'stream' ) );
-			return;
```
Comment (szepeviktor) [inline includes/wp-cli.php]: An empty table is the usualy wp-cli output.

File: includes/wp-cli.php [hunk 8]
```diff
@@ -8,96 +8,152 @@
...
+		// make structure Formatter compatible
+		foreach ( $records as $key => $record ) {
+			$formatted_records[$key] = array();
+			foreach ( $record as $field_name => $field ) {
+				$formatted_records[$key] = array_merge( $formatted_records[$key], $this->any2array( $field_name, $field ) );
+			}
 		}
 
-		$fields = array_map( 'trim', explode( ',', $query_args['fields'] ) );
```
Comment (szepeviktor) [inline includes/wp-cli.php]: on the comand line, it is very hard to append a `SPACE` character

File: includes/wp-cli.php [hunk 9]
```diff
@@ -8,96 +8,152 @@
...
-
-				array_walk( $values, function( &$value ) {
-					$value = is_array( $value ) ? $value[0] : $value;
-				});
-
-				$longest = max( array_map( 'strlen', $values ) );
-				$value   = is_array( $record->$field ) ? implode( ', ', $record->$field ) : $record->$field;
-				$output .= $value;
-				$diff    = absint( $longest - strlen( $value ) );
+		$formatter = new \WP_CLI\Formatter(
```
Comment (szepeviktor) [inline includes/wp-cli.php]: wp-cli way of displaying output

File: includes/wp-cli.php [hunk 10]
```diff
@@ -8,96 +8,152 @@
...
-			WP_CLI::line( trim( $output ) );
+		} elseif ( is_array( $object ) ) {
+			$array[$name] = $object[0];
+		} else {
+			$array[$name] = $object;
 		}
-
-		$found   = count( $records );
-		$stop    = microtime( true ) - $start;
-		$message = sprintf( _n( '1 result found in %s seconds.', '%s results found in %s seconds.', $found, 'stream' ), number_format( $found ), number_format( $stop, 4 ) );
```
Comment (szepeviktor) [inline includes/wp-cli.php]: the timer distracts the user

File: includes/wp-cli.php [hunk 11]
```diff
@@ -8,96 +8,152 @@
...
-
-		WP_CLI::success( $message );
```
Comment (szepeviktor) [inline includes/wp-cli.php]: success should detected by the return value as `echo $?` - we are on the command line
a lot of shell prompts include this value or turns red when not zero

File: includes/wp-cli.php [hunk 12]
```diff
@@ -8,96 +8,152 @@
...
 	}
 
 	/**
-	 * Checks for a Stream connection and displays an error or success message
+	 * Checks for a Stream connection and displays an error or success message.
 	 *
 	 * @return void
 	 */
-	private static function connection() {
-		WP_CLI::line( __( 'Establishing secure connection with WP Stream...', 'stream' ) );
```
Comment (szepeviktor) [inline includes/wp-cli.php]: this prevents processing of the output, e.g. by `grep`

File: includes/wp-cli.php [hunk 13]
```diff
@@ -8,96 +8,152 @@
...
-
+	private function connection() {
 		$query = wp_stream_query( array( 'records_per_page' => 1, 'fields' => 'created' ) );
 
 		if ( ! $query ) {
 			WP_CLI::error( __( 'SITE IS DISCONNECTED', 'stream' ) );
```
Comment (szepeviktor) [inline includes/wp-cli.php]: this is necessary

**Comment by szepeviktor** [comment]
This must have an least one error. Please review.

**Comment by szepeviktor** [comment]
Has a very nice output. Do you need csv, json and *count* output?

```
+------------------------------+----------------+--------+------------------------+---------------+------------------------------------+
| created                      | ip             | author | author_meta.user_login | author_role   | summary                            |
+------------------------------+----------------+--------+------------------------+---------------+------------------------------------+
| 2015-02-25T18:07:20.709+0000 | 192.168.12.146 | 1      | viktor                 | administrator | "1" post published                 |
| 2015-02-25T18:07:18.045+0000 | 192.168.12.146 | 1      | viktor                 | administrator | "1" post drafted                   |
| 2015-02-25T18:04:51.063+0000 | 192.168.12.146 | 1      | viktor                 | administrator | "What's running" plugin activated  |
| 2015-02-25T18:04:01.200+0000 | 192.168.12.146 | 1      | viktor                 | administrator | Site connected to Stream           |
+------------------------------+----------------+--------+------------------------+---------------+------------------------------------+
```

**Comment by szepeviktor** [comment]
This turned out to be a radical PR.

File: includes/wp-cli.php (hunk 5 above)
Comment (fjarrett) [inline includes/wp-cli.php]: Hey @szepeviktor, sorry for my absence - I just returned from paternity leave. Yeah, I think a separate wiki page for WP-CLI makes sense for that. Just created this placeholder: https://github.com/wp-stream/stream/wiki/WP-CLI-Command

**Comment by fjarrett** [comment]
@szepeviktor Just a few CI errors to take note of: https://travis-ci.org/wp-stream/stream/jobs/52189700#L890

**Comment by szepeviktor** [comment]
Was it CodeSniffer?

File: includes/wp-cli.php (hunk 5 above)
Comment (szepeviktor) [inline includes/wp-cli.php]: OK?
https://github.com/wp-stream/stream/wiki/WP-CLI-Command

**Comment by szepeviktor** [comment]
I've done it for free. Could you give me a trial PRO lincence?

**Comment by szepeviktor** [comment]
All green.

**Comment by fjarrett** [comment]
Thanks @szepeviktor. We are using the [WordPress Coding Standards](https://github.com/WordPress-Coding-Standards/WordPress-Coding-Standards) project to enforce coding style, which uses [PHP_CodeSniffer](http://pear.php.net/package/PHP_CodeSniffer/redirected).

Absolutely, please e-mail me your Stream Site ID frankie@wp-stream.com

File: includes/wp-cli.php (hunk 4 above)
Comment (fjarrett) [inline includes/wp-cli.php]: @szepeviktor Fair enough. Better results can be seen with the `time` command anyway.

`$ time wp stream query`

File: includes/wp-cli.php (hunk 5 above)
Comment (fjarrett) [inline includes/wp-cli.php]: @szepeviktor Nice!

Comment (szepeviktor) [inline includes/wp-cli.php]: Clear to merge?

**Comment by fjarrett** [comment]
@szepeviktor Nicely done! Thanks for these many improvements. Please do shoot me an email when you can.

File: includes/wp-cli.php (hunk 10 above)
Comment (fjarrett) [inline includes/wp-cli.php]: I added the results count back in, it's helpful meta to know when dealing with large result sets.

https://github.com/wp-stream/stream/commit/1377185f004fac5486fba3b749488889be948312

**Pull request merged**